File: pkg/dart2cpp/sample/dart/test_operator_equals.dart
// 测试 operator== 中的类型推断

class Vector {
  final double x, y;

  Vector(this.x, this.y);

  @override
  bool operator ==(Object other) =>
      other is Vector && x == other.x && y == other.y;

  @override
  int get hashCode => Object.hash(x, y);
}

void main() {
  var v1 = Vector(1.0, 2.0);
  var v2 = Vector(1.0, 2.0);
  var v3 = Vector(3.0, 4.0);

  print('v1 == v2: ${v1 == v2}');
  print('v1 == v3: ${v1 == v3}');
}

File: pkg/dart2cpp/sample/dart/test_operator_equals.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: Vector
// ============================================================================

class Vector {
public:
  Double x;
  Double y;
  Vector(Double x, Double y) : x(x), y(y) {
  }
  
  Bool operator_equals(ObjectPtr<Object> other) {
    return dart_is<ObjectPtr<Vector>>(other) && (this->x == dart_cast<ObjectPtr<Vector>>(other)->x) && (this->y == dart_cast<ObjectPtr<Vector>>(other)->y);
  }
  
  Int hashCode() {
    return Object::hash(this->x, this->y);
  }
  
};

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    auto v1 = ObjectPtr<Vector>(new Vector(dart_double(1.0), dart_double(2.0)));
auto v2 = ObjectPtr<Vector>(new Vector(dart_double(1.0), dart_double(2.0)));
auto v3 = ObjectPtr<Vector>(new Vector(dart_double(3.0), dart_double(4.0)));
dart_print(dart_string("v1 == v2: ") + ((v1 == v2)).toString());
dart_print(dart_string("v1 == v3: ") + ((v1 == v3)).toString());
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
